### PR TITLE
[stag] PLT-1171 Mandatory connectionQualifiedName for CU Persona metadata/data policy

### PR DESCRIPTION
**Pull request by nikhilbonte21**

### CODE REVIEW

File: auth-agents-common/src/main/java/org/apache/atlas/policytransformer/PersonaCachePolicyTransformer.java [hunk 1]
```diff
@@ -128,16 +129,22 @@ public List<AtlasEntityHeader> transform(AtlasEntityHeader atlasPolicy) {
...
         if (assets.size() == 1) {
-            AtlasEntity connection;
-            try {
-                connection = getConnectionEntity(entityRetriever, assets.get(0));
-            } catch (AtlasBaseException abe) {
-                return false;
+            String connQNAttr = getPolicyConnectionQN(atlasPolicy);
+
+            if (StringUtils.isNotEmpty(connQNAttr)) {
+                return connQNAttr.equals(assets.get(0));
```
Comment (mehtaanshul) [inline auth-agents-common/src/main/java/org/apache/atlas/policytransformer/PersonaCachePolicyTransformer.java:138]: add verification here as well - if connection exists or not

Comment (mehtaanshul) [reply auth-agents-common/src/main/java/org/apache/atlas/policytransformer/PersonaCachePolicyTransformer.java:138]: what is not equal to assets.get(0)

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java [hunk 2]
```diff
@@ -156,26 +138,36 @@ public void validate(AtlasEntity policy, AtlasEntity existingPolicy,
                     validateParam (!PERSONA_POLICY_VALID_SUB_CATEGORIES.contains(policySubCategory),
                             "Please provide valid value for attribute " + ATTR_POLICY_SUB_CATEGORY + ":"+ PERSONA_POLICY_VALID_SUB_CATEGORIES);
 
+
+                    List<String> resources = getPolicyResources(policy);
+                    validateParam (CollectionUtils.isEmpty(resources), "Please provide attribute " + ATTR_POLICY_RESOURCES);
+
+                    //validate persona policy resources keys
+                    Set<String> policyResourceKeys = resources.stream().map(x -> x.split(RESOURCES_SPLITTER)[0]).collect(Collectors.toSet());
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:146]: x.split(RESOURCES_SPLITTER)[0] - use `:`

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java [hunk 3]
```diff
@@ -156,26 +138,36 @@ public void validate(AtlasEntity policy, AtlasEntity existingPolicy,
                     validateParam (!PERSONA_POLICY_VALID_SUB_CATEGORIES.contains(policySubCategory),
                             "Please provide valid value for attribute " + ATTR_POLICY_SUB_CATEGORY + ":"+ PERSONA_POLICY_VALID_SUB_CATEGORIES);
 
+
+                    List<String> resources = getPolicyResources(policy);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:142]: Why empty resources should not be allowed?

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java [hunk 4]
```diff
@@ -156,26 +138,36 @@ public void validate(AtlasEntity policy, AtlasEntity existingPolicy,
...
+                    validateParam(CollectionUtils.isNotEmpty(policyResourceKeys),
+                            "Please provide valid type of policy resources: " + PERSONA_POLICY_VALID_RESOURCE_KEYS);
+
+
                     if (POLICY_SUB_CATEGORY_DATA.equals(policySubCategory)) {
                         validateParam (!POLICY_RESOURCE_CATEGORY_PERSONA_ENTITY.equals(getPolicyResourceCategory(policy)), "Invalid resource category for Persona");
-                    } else {
+                        validateConnection(getPolicyConnectionQN(policy), resources);
+
+                    } else if (POLICY_SUB_CATEGORY_METADATA.equals(policySubCategory)) {
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:156]: discuss this

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java [hunk 5]
```diff
@@ -267,7 +271,27 @@ public void validate(AtlasEntity policy, AtlasEntity existingPolicy,
         }
     }
 
-    private static void validateParentUpdate(AtlasEntity policy, AtlasEntity existingPolicy) throws AtlasBaseException {
+    private void validateEntityResources(String connQn, List<String> resources) throws AtlasBaseException {
+        List<String> entityResources = getFilteredPolicyResources(resources, RESOURCES_ENTITY);
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:275]: does value of resource entity include `:`

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java [hunk 6]
```diff
@@ -223,22 +215,34 @@ public void validate(AtlasEntity policy, AtlasEntity existingPolicy,
                 String policySubCategory = policy.hasAttribute(ATTR_POLICY_SUB_CATEGORY) ? getPolicySubCategory(policy) : getPolicySubCategory(existingPolicy);
 
                 if (POLICY_CATEGORY_PERSONA.equals(policyCategory)) {
-                    validateParam (policy.hasAttribute(ATTR_POLICY_RESOURCES) && CollectionUtils.isEmpty(getPolicyResources(policy)),
+                    List<String> resources = getPolicyResources(policy);
+                    validateParam (policy.hasAttribute(ATTR_POLICY_RESOURCES) && CollectionUtils.isEmpty(resources),
                             "Please provide attribute " + ATTR_POLICY_RESOURCES);
 
+                    //validate persona policy resources keys
+                    Set<String> policyResourceKeys = resources.stream().map(x -> x.split(RESOURCES_SPLITTER)[0]).collect(Collectors.toSet());
```
Comment (mehtaanshul) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:223]: use `:`

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java (hunk 3 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:142]: While creating policies we need policy resources

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java (hunk 5 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:275]: Yes

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java (hunk 6 above)
Comment (nikhilbonte21) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/AuthPolicyValidator.java:223]: public static final String RESOURCES_SPLITTER = ":";

**Pull request merged**